versioned/persist/tests/src/main/java/org/projectnessie/versioned/persist/tests/DatabaseAdapterExtension.java

```diff
@@ -0,0 +1,194 @@
+  }
+
+  private static class NoopConfigurer implements TestConfigurer {
+    @Override
+    public <C extends DatabaseAdapterConfig> DatabaseAdapterConfig configure(C config) {
+      return config;
+    }
+  }
+
+  protected static <C extends DatabaseAdapterConfig> DatabaseAdapter createAdapter(
+      String adapterName,
+      TestConfigurer testConfigurer,
+      Function<C, DatabaseAdapterConfig> configurer) {
+    return createAdapter(
+        testConfigurer,
+        DatabaseAdapterFactory.<C>loadFactoryByName(adapterName).newBuilder(),
+        configurer);
+  }
+
+  protected static <C extends DatabaseAdapterConfig> DatabaseAdapter createAdapter(
+      TestConfigurer testConfigurer,
+      DatabaseAdapterFactory.Builder<C> adapterBuilder,
+      Function<C, DatabaseAdapterConfig> configurer) {
+    adapterBuilder =
+        adapterBuilder
+            .configure(SystemPropertiesConfigurer::configureFromSystemProperties)
+            // default to a quite small max-size for the CommitLogEntry.keyList + KeyListEntity
+            // This is necessary for AbstractManyKeys to work properly!!
+            .configure(c -> c.withMaxKeyListSize(2048))
+            .configure(configurer)
+            .configure(testConfigurer::configure);
+
+    return adapterBuilder.build();
+  }
+
+  class AdapterResource implements CloseableResource {
+    private final DatabaseAdapter databaseAdapter;
+
+    AdapterResource(DatabaseAdapter databaseAdapter) {
+      this.databaseAdapter = databaseAdapter;
+    }
+
+    @Override
+    public void close() throws Throwable {
+      databaseAdapter.close();
+
+      afterCloseAdapter();
+    }
+  }
+
+  @Override
+  public void beforeAll(ExtensionContext context) throws Exception {
+    Class<?> testClass = context.getRequiredTestClass();
+    findAnnotatedFields(testClass, Adapter.class, ReflectionUtils::isStatic)
+        .forEach(
+            field -> {
+              assertValidStaticFieldCandidate(field);
+              Adapter annotation =
+                  findAnnotation(field, Adapter.class).orElseThrow(IllegalStateException::new);
+              try {
+                if (context.getStore(NAMESPACE).get(KEY) != null) {
+                  throw new JUnitException(
+                      "Only one instance of '@DatabaseAdapterExtension.Adapter static DatabaseAdapter databaseAdapter;'"
+                          + "field allowed per test class");
+                }
+
+                TestConfigurer configurer =
+                    makeAccessible(annotation.configurer().getDeclaredConstructor()).newInstance();
+                DatabaseAdapter adapter =
+                    context
+                        .getStore(NAMESPACE)
+                        .getOrComputeIfAbsent(
+                            KEY,
+                            key -> new AdapterResource(createAdapter(context, configurer)),
+                            AdapterResource.class)
```

Comment:
This looks like we will be creating and closing adapters and related storage databases per test class, if I'm not mistaken. Is that intended? Currently we create one storage for all adapter tests.

Reply:
No, it's created for the whole "adapter test class" including all subclasses